File: 003_longest_substring_without_repeating_characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        s1 = ""
        if len(s) == 1:
            return 1
        length = 0
        maximum = 0
        for i in s:
            if i in s1:
                if maximum < length:
                    maximum = length
                length = 0
                s1 = ""
                s1 += i
                length += 1
            else:
                s1 += i
                length += 1
                print(s1)
        return maximum

File: 003_longest_substring_without_repeating_characters.cpp
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        vector<int> lastIndex(256, -1); // stores last seen index of each ASCII char
        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < s.size(); right++)
        {
            char c = s[right];

            // If we've seen this char and it's inside the current window
            if (lastIndex[c] >= left)
            {
                left = lastIndex[c] + 1; // shrink window
            }

            lastIndex[c] = right; // update last seen index
            maxLen = max(maxLen, right - left + 1);
        }

        return maxLen;
    }
};
